Automattic/wp-calypso
# Credit Card forms: Add checkbox that assigns new cards to all subscriptions

#### Changes proposed in this Pull Request

This adds a "Use this payment method for all subscriptions on my account." checkbox to the Stripe credit card forms in checkout and payment method management. A checkbox for this (with slightly different wording) already exists in the non-subscription "Add Payment Method" form, but this removes it in favor of the new one.

<img width="485" alt="Screen Shot 2021-10-01 at 7 15 24 PM" src="https://user-images.githubusercontent.com/2036909/135695332-1ed278b6-2097-42a5-94d7-6adc274d380b.png">

On the non-subscription "Add Payment Method" form, the checkbox will be pre-checked (as it is currently), but all other instances of the checkbox (checkout and subscription-level Add/Change Payment Method forms) will default to unchecked.

If the new element is checked when each of the relevant forms is submitted, it will replace the payment method on _all_ of the user's active subscriptions with the newly added credit card.

For simplicity during implementation, this functionality is not yet being added to existing credit cards, Ebanx credit cards, PayPal, or any other payment method, but it may be added (particularly to PayPal) in the future.

Requires D67859-code to update the endpoints to support the new property. 

#### Motivation

The reason for this change is that many users come to these forms intending to update an expiring or otherwise incorrect payment method, but may be unaware that they need to make the change to each of their subscriptions. As a result they may update an expiring card on one subscription and then be confused when they receive a notification that another subscription is still using the expiring card.

This is quite common since users often purchase several products at once (e.g.: a plan and a domain) and are not clear that the payment methods for these purchases must be managed individually. While a more advanced payment management interface would be nice, it is our hope that while that project is being discussed, this checkbox will allow such users to update their card once and therefore avoid some confusion.

Part of 509-gh-Automattic/payments-shilling

#### To do

- [x] Add checkbox to all Stripe card forms.
- [x] Remove existing checkbox on add payment method form.
- [x] Send data to payment method add endpoint in place of old checkbox data.
- [x] Send data to transactions endpoint.
- [x] Add real "Learn more" link target to checkbox label.
- [x] Make checkbox pre-checked on non-subscription add payment method form.
- [x] Hide checkbox for non-wpcom purchases.
- [x] Modify transaction endpoint to use the new property.
- [x] Remove CSS for removed checkbox.

#### Screenshots

Here's the old checkbox in the non-subscription "Add Payment Method" form that is being replaced by this PR:

<img width="581" alt="Screen Shot 2021-10-01 at 7 26 58 PM" src="https://user-images.githubusercontent.com/2036909/135695650-f9d23fcf-19d1-4bc3-8735-6af2a25e88fe.png">

Here's how the new checkbox looks on the "Add Payment Method" form:

<img width="584" alt="Screen Shot 2021-10-01 at 7 28 54 PM" src="https://user-images.githubusercontent.com/2036909/135695734-42b5ac61-47c3-4d10-bbf9-ba0b60924563.png">

Here's how the new checkbox looks in checkout:

<img width="502" alt="Screen Shot 2021-10-01 at 4 17 13 PM" src="https://user-images.githubusercontent.com/2036909/135682261-60dca421-8402-41b6-bfd5-1bdacc9794e9.png">

#### Testing instructions

To test checkout:

- Use a site that has an existing wpcom purchase (eg: a plan or a domain).
- Add a new product to the cart (if your existing purchase is a plan, this new product cannot also be a plan because it will replace the old one) and visit checkout.
- Fill out the checkout form and select "Credit or Debit card" as your payment method.
- Check the "Use this payment method for all subscriptions on my account." checkbox.
- Click to submit the purchase.
- Visit `/me/purchases` and verify that your existing purchase has been changed to use the new card as its payment method.

To test subscription payment method management:

- First, use a site that has an existing non-wpcom purchase (eg: a Jetpack site).
- Visit `/me/purchases`, select your existing purchase, click "Change payment method", and verify that you _do not_ see the "Use this payment method for all subscriptions on my account." checkbox in the new card form.
- Next, use a site has _more than one_ existing wpcom purchase (eg: a plan or a domain).
- Visit `/me/purchases`, select your existing purchase, and click "Change payment method". 
- Select "Credit or Debit card", fill in the new card form and check the "Use this payment method for all subscriptions on my account." checkbox.
- Click to submit the form.
- Visit `/me/purchases` and verify that all your purchases have been changed to use the new card as their payment method.

To test non-subscription payment method management:

- Use a site has _more than one_ existing wpcom purchase (eg: a plan or a domain).
- Visit `/me/purchases/add-payment-method`.
- Fill in the new card form and verify that the "Use this payment method for all subscriptions on my account." checkbox is pre-checked.
- Click to submit the form.
- Visit `/me/purchases` and verify that all your purchases have been changed to use the new card as their payment method.
